### PR TITLE
better up / down handling in FilterList

### PR DESCRIPTION
**Pull request by kjk**
As described in https://github.com/silverbulletmd/silverbullet/issues/230

I tried routing keydown via CodeMIrror / MiniEditor like keyup but couldn't figure it out.

Instead I hooked global keydown.

### CODE REVIEW

**Comment by zefhemel** [comment]
Thanks I'll have a closer look at this a bit later (holiday prep going on here). Key handling is a pain in the ass, so I need to do some proper testing to avoid any regressions.

File: web/components/filter.tsx [hunk 1]
```diff
@@ -126,6 +126,56 @@ export function FilterList({
     };
   }, []);
 
+  useEffect(() => {
+    // TODO: why is this called after each setSelectionOption
```
Comment (zefhemel) [inline web/components/filter.tsx:130]: Spoiler: because of line 177

File: web/components/filter.tsx [hunk 2]
```diff
@@ -126,6 +126,56 @@ export function FilterList({
...
+      return true;
+    }
+
+    document.addEventListener("keydown", onKeyDown);
+
+    return () => {
+      document.removeEventListener("keydown", onKeyDown);
+    };
+
+  }, [selectedOption, matchingOptions]);
```
Comment (zefhemel) [inline web/components/filter.tsx:177]:
```suggestion
  }, []);
```

to do this just once?

**Comment by zefhemel** [comment]
I think this shouldn’t happen through a global event listener, instead the MiniEditor should be expanded with an `onKeyDown` event handler probably.

File: web/components/filter.tsx (hunk 1 above)
Comment (kjk) [reply web/components/filter.tsx:130]: I'm totally React newbie and yeah, I figured out that those are dependencies and when selectOption changes it causes that function to re-run.

The thing is: it shouldn't be necessary to re-bind event handler on every state change, only when the element mounts. I had no problem getting that in Svelte.

However, if I don't pass selectOption as dependency then it somehow gets memoized and only the first modification takes effect.

The docs hint that state changes must happen inside element but I don't know how to push that code there.

Like I said: React newbie. Naively there should be a way to only bind event handler on mount and still modify state from handler but didn't figure out how.

**Comment by kjk** [comment]
I tried that as my first approach (i.e. added onkeydown to MiniEditor and hooked it up to MiniEditor and FilterList, same as onkeyup) but it straight up didn't work i.e. onkeydown handler in Code Mirror wasn't called. 

Also: a remaining mystery is why up / down also affects Code Mirror (moves cursor to start/end of field) even when I stop propagation after handling up/down keydown.

**Comment by zefhemel** [comment]
@kjk sorry for butting into your PR 😆  Didn't realize I could just push commits here. What I did is basically move your code into the mini editor component so that other places can use a similar mechanism. Indeed, you were right, CM doesn't propagate the keydown event so you have to capture it at a higher level.

**Pull request merged**